Runtime/Scripts/RGBotServerListener.cs

```diff
@@ -515,41 +516,41 @@ private void FixedUpdate()
         }
 
         /**
-         * Gets the entire game state by searching for all RGState game objects and gather their
+         * Gets the entire game state by searching for all RGEntity game objects and gathering their
          * states.
          */
         private Dictionary<string, IRGStateEntity> GetGameState()
         {
             var overlayAgent = this.gameObject.GetComponent<RGEntity>();
-            var statefulObjects = FindObjectsOfType<MonoBehaviour>(true).OfType<IRGState>();
+            
+            var statefulRGEntities = FindObjectsOfType<MonoBehaviour>(true).OfType<RGEntity>();
             var fullGameState = new Dictionary<string, IRGStateEntity>();
 
             // SADLY... Unity's threading model sucks and accessing the transform of an object must be done on the main thread only
             // thus, this code cannot really be run in parallel, causing a major object count scaling issue....
-            foreach (var statefulObject in statefulObjects)
+            foreach (var statefulRGEntity in statefulRGEntities)
             {
-                var gameObjectState = statefulObject.GetGameObjectState();
+                // Give the Entity its 'core' state fields whether their are custom RGstates on the object or not
+                // Custom states can then override these values
+                var entityCoreState = RGState.GenerateCoreStateForRGEntity(statefulRGEntity);
+                fullGameState[entityCoreState.id.ToString()] = entityCoreState;
                 
-                // If user implements their own state script that implements IRGState
-                // then we won't be able to cast it to RGState.
-                // We'll still try to add it to the fullGameState...
-                // but we can't access its RGEntity component :-(
-                var rgState = statefulObject as RGState;
-                if (rgState != null) 
-                {
-                    // GameObjects must have an RGEntity to be tracked in the game state.
-                    var rgEntity = rgState.GetComponentInParent<RGEntity>();
-                    if (rgEntity == null) continue;
+                // get the state behaviors on this game object
+                var rgStates = statefulRGEntity.gameObject.GetComponents<IRGState>();
                 
-                    if (true.Equals(gameObjectState.isPlayer))
+                // build up the state for this entity from all the different state classes
+                foreach (var rgState in rgStates)
+                {
+                    var gameObjectStateEntity = rgState.GetGameObjectState();
+                    if (true.Equals(gameObjectStateEntity.isPlayer))
```

Comment:
`isPlayer` check here should be from the core (RGEntity) state, and we don't need to repeat it for every RGstate script

Reply:
code updated / resolved